File: Practice/DFS_BFS/q15/q15.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>

using namespace std;

vector<int> ans;
vector<int> cities[300001];
bool chk[300001];
int n, m, k, x;

void bfs(int x){
    chk[x] = true;
    queue<pair<int,int>> q;
    q.push(make_pair(x, 0));
    while(!q.empty()){
        vector<int> city = cities[q.front().first];
        int dist = q.front().second;
        if(dist == k) ans.push_back(q.front().first);
        q.pop();
        for(int i = 0; i < city.size(); i++){
            int to = city[i];
            if(!chk[to]){
                chk[to] = true;
                q.push(make_pair(to, dist+1));
            }
        }
    }
}

int main(void){
    cin >> n >> m >> k >> x;
    for(int i = 0; i < m; i++){
        int a;
        int b;
        cin >> a >> b;
        cities[a].push_back(b);
    }

    bfs(x);

    if(ans.size() != 0){
        sort(ans.begin(), ans.end());
        for(int i = 0; i < ans.size(); i++){
            cout << ans[i] << "\n";
        }
    }else{
        cout << -1 << "\n";
    }
}

File: Practice/DFS_BFS/q15/q15.py
from collections import deque

n, m, k, x = map(int, input().split())
adj = [[] for _ in range(n+1)]
chk = [False for _ in range(n+1)]
ans = []
for _ in range(m):
    c1, c2 = map(int, input().split())
    adj[c1].append(c2)

def get_k_dist_city(start):
    queue = deque()
    queue.append((start, 0))
    chk[start] = True
    while len(queue) != 0:
        city = queue.popleft()
        if city[1] == k:
            ans.append(city[0])
        for c in adj[city[0]]:
            if chk[c] == False:
                queue.append((c, city[1]+1))
                chk[c] = True

get_k_dist_city(x)
if len(ans) == 0:
    print(-1)
else:
    ans.sort()
    for c in ans:
        print(c)